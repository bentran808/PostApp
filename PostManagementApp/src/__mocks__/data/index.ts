const initialUser = {
  id: '',
  avatar: '',
  name: '',
  gender: '',
  email: '',
  phone: '',
  address: '',
  role: ''
};

const mockAdmin = {
  id: '621c4188dece66a40dad2951',
  avatar: 'https://picsum.photos/200',
  name: 'Admin',
  gender: 'male',
  email: 'admin@admin.com',
  phone: '+84 (919) 507231',
  address: '199 Bergen Street, Bowie, Oregon, 7933',
  role: 'admin'
};

const mockUser = {
  id: '621c41883112f1c216433dcf',
  avatar: 'https://picsum.photos/200',
  name: 'Mcintyre Morrison',
  gender: 'male',
  email: 'test1@email.com',
  phone: '+84 (838) 539391',
  address: '313 Menahan Street, Aurora, North Dakota, 9139',
  role: 'user'
};

const mockPost = {
  author: mockAdmin,
  company: 'bb',
  year: 2222,
  type: 'b',
  status: true,
  price: 23333111,
  address: '199 Bergen Street, Bowie, Oregon, 7933',
  title: 'bb update',
  description: 'bb',
  photos: [
    {
      height: 480,
      uri: 'file:///data/user/0/com.PostManagementApp/cache/rn_image_picker_lib_temp_283c7c3b-b014-4ef4-b95a-c092bcefe2aa.jpg',
      width: 480,
      fileName: 'rn_image_picker_lib_temp_283c7c3b-b014-4ef4-b95a-c092bcefe2aa.jpg',
      type: 'image/jpeg',
      fileSize: 6689
    },
    {
      height: 480,
      uri: 'file:///data/user/0/com.PostManagementApp/cache/rn_image_picker_lib_temp_283c7c3b-b014-4ef4-b95a-c092bcefe2aa.jpg',
      width: 480,
      fileName: 'rn_image_picker_lib_temp_283c7c3b-b014-4ef4-b95a-c092bcefe2aa.jpg',
      type: 'image/jpeg',
      fileSize: 6689
    }
  ],
  pending: false,
  createdAt: 1647507076963,
  updatedAt: 1647511146749,
  id: 2
};

const mockLikes = [
  {
    author: mockAdmin,
    postId: 2,
    createdAt: 1647509052801,
    updatedAt: 1647509052801,
    id: 1
  }
];

const mockCommentsList = [
  {
    author: mockAdmin,
    postId: 2,
    content: 'test',
    createdAt: 1647509052801,
    updatedAt: 1647509052801,
    id: 1
  },
  {
    author: {
      id: '621c4188e8365653f9326450',
      avatar: 'https://picsum.photos/200',
      name: 'Frost Spence',
      gender: 'male',
      email: 'frostspence@entogrok.com',
      phone: '+84 (859) 508289',
      address: '745 Blake Avenue, Belmont, District Of Columbia, 4099',
      role: 'user'
    },
    postId: 2,
    content: 'test2',
    createdAt: 1647509052801,
    updatedAt: 0,
    id: 2
  }
];

const mockNewPost = {
  author: mockAdmin,
  company: 'qwe',
  year: 2021,
  type: 'qwe',
  status: true,
  price: 10000,
  address: 'qwe',
  title: 'qwe',
  description: 'qwe',
  photos: [],
  pending: true
};

const mockEditedPost = {
  '2': mockPost,
  '3': {
    id: 3,
    ...mockNewPost
  }
};

const mockApprovedPost = {
  '2': mockPost,
  '3': {
    id: 3,
    ...mockNewPost,
    pending: false
  }
};

const mockLike = {
  author: mockAdmin,
  postId: 2,
  createdAt: 1647509052801,
  updatedAt: 1647509052801,
  id: 2
};

const mockNewComment = {
  author: mockAdmin,
  postId: 2,
  content: 'abc',
  createdAt: 1647509052801,
  updatedAt: 1647509052801,
  id: 3
};

const mockPhotos = [
  {
    fileName: 'rn_image_picker_lib_temp_283c7c3b-b014-4ef4-b95a-c092bcefe2aa.jpg',
    fileSize: 6689,
    height: 480,
    type: 'image/jpeg',
    uri: 'file:///data/user/0/com.PostManagementApp/cache/rn_image_picker_lib_temp_283c7c3b-b014-4ef4-b95a-c092bcefe2aa.jpg',
    width: 480
  },
  {
    fileName: 'rn_image_picker_lib_temp_283c7c3b-b014-4ef4-b95a-c092bcefe2aa.jpg',
    fileSize: 6689,
    height: 480,
    type: 'image/jpeg',
    uri: 'file:///data/user/0/com.PostManagementApp/cache/rn_image_picker_lib_temp_283c7c3b-b014-4ef4-b95a-c092bcefe2aa.jpg',
    width: 480
  }
];

export {
  initialUser,
  mockAdmin,
  mockUser,
  mockPost,
  mockLikes,
  mockCommentsList,
  mockNewPost,
  mockEditedPost,
  mockApprovedPost,
  mockLike,
  mockNewComment,
  mockPhotos
};
