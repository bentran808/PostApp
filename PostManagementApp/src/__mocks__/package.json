{
    "name": "__mock__",
    "private": true
}