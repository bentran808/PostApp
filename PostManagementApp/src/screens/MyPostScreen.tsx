import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

const MyPostScreen = () => (
    <View style={styles.body}>
        <Text style={styles.text}>My Post Tab</Text>
    </View>
);

const styles = StyleSheet.create({
    body: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center'
    },
    text: {
        fontSize: 30,
        fontWeight: 'bold'
    }
});

export default MyPostScreen;
