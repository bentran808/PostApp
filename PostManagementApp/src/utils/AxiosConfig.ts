import axios from 'axios';

export const axiosInstance = axios.create({
    baseURL: 'http://192.168.56.1:3000/',
    headers: {
        'Content-Type': 'application/json'
    }
});
