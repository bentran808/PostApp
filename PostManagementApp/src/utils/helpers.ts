import AsyncStorage from '@react-native-async-storage/async-storage';

const sortDesc = (arr: Post[] = []) => {
  return arr.sort((a, b) => ((a.id || 0) < (b.id || 0) ? 1 : -1));
};

const formatPrice = (price: Number) => {
  return price.toLocaleString('it-IT', {
    style: 'currency',
    currency: 'VND'
  });
};

const getApprovedPosts = (arr: Post[] = []) => arr.filter((item) => !item.pending);

const setAccessToken = async (access_token: string) =>
  await AsyncStorage.setItem('access_token', access_token);

const getAccessToken = async () => await AsyncStorage.getItem('access_token');

const removeAccessToken = async () => await AsyncStorage.removeItem('access_token');

export {
  sortDesc,
  formatPrice,
  getApprovedPosts,
  setAccessToken,
  getAccessToken,
  removeAccessToken
};
