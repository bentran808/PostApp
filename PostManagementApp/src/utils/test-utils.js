import React from 'react';
import { Provider } from 'react-redux';
import { Provider as PaperProvider } from 'react-native-paper';
import { render } from '@testing-library/react-native';
import createSagaMiddleware from 'redux-saga';
import { configureStore } from '@reduxjs/toolkit';
import authReducer from 'redux/slices/authSlice';
import postReducer from 'redux/slices/postSlices';

const sagaMiddleware = createSagaMiddleware();
const AllTheProviders = ({ children }) => {
  const store = configureStore({
    reducer: {
      auth: authReducer,
      post: postReducer
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        serializableCheck: false
      }).concat(sagaMiddleware)
  });
  return (
    <Provider store={store}>
      <PaperProvider>{children}</PaperProvider>
    </Provider>
  );
};

const customRender = (ui, options) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export { customRender as render };
