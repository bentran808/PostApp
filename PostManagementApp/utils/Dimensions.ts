import {Dimensions} from 'react-native';

const windowHeight = Dimensions.get('window').height;

export {windowHeight};
