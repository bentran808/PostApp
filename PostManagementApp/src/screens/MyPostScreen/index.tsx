/* eslint-disable react-hooks/exhaustive-deps */
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import axios from 'axios';
import React, { useCallback, useContext, useEffect, useState } from 'react';
import { FlatList, StyleSheet, View } from 'react-native';
import { Colors, Text } from 'react-native-paper';
import { axiosInstance } from '../../api';
import PostCard from '../../components/PostCard';
import { AppContext } from '../../navigation/AppContext';
import { sortDesc } from '../../utils/helpers';

type MyPostNavigationProp = NativeStackNavigationProp<RootStackParamList>;

interface MyPostProps {
    navigation: MyPostNavigationProp;
    route: {
        params: {
            data: Post;
        };
    };
}

const MyPostScreen = ({ navigation, route }: MyPostProps) => {
    const [posts, setPosts] = useState<Post[]>([]);
    const [loading, setLoading] = useState(false);
    const { user } = useContext(AppContext);
    const config = {
        headers: {
            Authorization: `Bearer ${user.access_token}`
        }
    };

    useEffect(() => {
        const source = axios.CancelToken.source();
        const getAllPosts = async () => {
            try {
                const response = await getPosts();
                if (response.status === 200) {
                    setPosts(sortDesc(getMyPosts(response.data)));
                    return;
                } else {
                    throw new Error('Failed to fetch users');
                }
            } catch (error) {
                if (axios.isCancel(error)) {
                    console.log('Data fetching cancelled');
                } else {
                    console.log(error);
                }
            }
        };
        getAllPosts();
        return () => source.cancel('Data fetching cancelled');
    }, [user, user.access_token]);

    useEffect(() => {
        const data = route.params && route.params.data;
        console.log('data', data);
        if (data) {
            if (data.createdAt === data.updatedAt) {
                setPosts(sortDesc(posts.concat(data)));
            } else {
                setPosts(sortDesc(posts.map((post) => (data.id === post.id ? data : post))));
            }
        }
    }, [route.params]);

    const getPosts = async () => await axiosInstance.get('api/posts', config);
    const getMyPosts = (arr: Post[]) =>
        user.data.role === 'admin'
            ? arr.filter((post) => post.pending)
            : arr.filter((post) => post.author.id === user.data.id);

    const onDelete = async (postId: Number) => {
        try {
            const response = await axiosInstance.delete(`api/posts/${postId}`, config);
            if (response.status === 200) {
                const newPosts = posts?.filter((item) => item.id !== postId);
                setPosts(newPosts);
                return;
            } else {
                throw new Error('Failed to delete a post');
            }
        } catch (error) {
            console.log(error);
        }
    };

    const onRefresh = useCallback(async () => {
        setLoading(true);
        try {
            const response = await getPosts();
            if (response.status === 200) {
                setPosts(sortDesc(getMyPosts(response.data)));
            } else {
                throw new Error('Failed to refresh posts');
            }
        } catch (error) {
            console.log(error);
        }
        setLoading(false);
    }, []);

    const approvePost = async (postId: Number) => {
        try {
            const response = await axiosInstance.patch(
                `api/posts/${postId}`,
                {
                    pending: false
                },
                {
                    headers: {
                        Authorization: `Bearer ${user.access_token}`
                    }
                }
            );
            if (response.status === 200) {
                setPosts(
                    getMyPosts(
                        posts.map((post) => (response.data.id === post.id ? response.data : post))
                    )
                );
                return;
            } else {
                throw new Error('Failed to approved a pending post');
            }
        } catch (error) {
            console.log(error);
        }
    };

    const rejectPost = async (postId: Number) => {
        try {
            const response = await axiosInstance.delete(`api/posts/${postId}`, config);
            if (response.status === 200) {
                const newPosts = posts?.filter((item) => item.id !== postId);
                setPosts(newPosts);
                return;
            } else {
                throw new Error('Failed to reject a pending post');
            }
        } catch (error) {
            console.log(error);
        }
    };

    return (
        <View style={styles.container}>
            <FlatList
                data={posts}
                refreshing={loading}
                onRefresh={onRefresh}
                renderItem={({ item }: { item: Post }) => (
                    <PostCard
                        item={item}
                        onPress={() =>
                            navigation.navigate('ShowImage', {
                                url: {
                                    uri: item.author.avatar
                                }
                            })
                        }
                        onDelete={() => onDelete(item.id)}
                        onEdit={() =>
                            navigation.navigate('AddPost', {
                                editedId: item.id
                            })
                        }
                        isMyPost={true}
                        approvePost={() => approvePost(item.id)}
                        rejectPost={() => rejectPost(item.id)}
                    />
                )}
                ListEmptyComponent={() => <Text>There are no posts to approve</Text>}
                showsVerticalScrollIndicator={false}
            />
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        alignItems: 'center',
        backgroundColor: Colors.white,
        padding: 20
    }
});

export default MyPostScreen;
