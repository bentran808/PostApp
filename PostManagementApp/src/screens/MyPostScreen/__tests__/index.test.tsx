import React from 'react';
import MyPostScreen from '..';
import { render } from '../../../utils/test-utils';
import { mockPost } from '../../../__mocks__/data';

describe('Test My Post Screen', () => {
  const route = {
    params: {
      data: mockPost
    }
  };
  let navigation: any;

  beforeEach(() => {
    navigation = {
      navigate: jest.fn()
    };
  });

  test('should render correctly', () => {
    const { toJSON } = render(<MyPostScreen navigation={navigation} route={route} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
