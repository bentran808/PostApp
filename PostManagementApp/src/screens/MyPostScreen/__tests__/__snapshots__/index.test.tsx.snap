// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test My Post Screen should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flex": 1,
        "padding": 20,
      }
    }
  >
    <RCTScrollView
      ListEmptyComponent={[Function]}
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onRefresh={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
        />
      }
      refreshing={false}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      showsVerticalScrollIndicator={false}
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <RCTRefreshControl />
      <View>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              undefined,
            ]
          }
        >
          There are no posts to approve
        </Text>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
